### ISSUE TITLE
[REVIEW]: IndeterminateBeam: A Python package for solving 1D indeterminate beams

### ISSUE THREAD

**Issue by whedon** [openjournals/jose-reviews#111]
**Submitting author:** @jessebonanno (<a href="http://orcid.org/0000-0002-4996-6813">Jesse Bonanno</a>)
**Repository:** <a href="https://github.com/JesseBonanno/IndeterminateBeam/" target ="_blank">https://github.com/JesseBonanno/IndeterminateBeam/</a>
**Version:** v2.1.6
**Editor:** @moorepants
**Reviewer:** @amcc1996 , @nicoguaro
**Archive:** <a href="https://doi.org/10.5281/zenodo.4851065" target="_blank">10.5281/zenodo.4851065</a>

## Status

[![status](https://jose.theoj.org/papers/1b9d108f658582001ce265c4b66811db/status.svg)](https://jose.theoj.org/papers/1b9d108f658582001ce265c4b66811db)

Status badge code:

```
HTML: <a href="https://jose.theoj.org/papers/1b9d108f658582001ce265c4b66811db"><img src="https://jose.theoj.org/papers/1b9d108f658582001ce265c4b66811db/status.svg"></a>
Markdown: [![status](https://jose.theoj.org/papers/1b9d108f658582001ce265c4b66811db/status.svg)](https://jose.theoj.org/papers/1b9d108f658582001ce265c4b66811db)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/JesseBonanno/IndeterminateBeam/" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@amcc1996  & @nicoguaro, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @moorepants know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSE conflict of interest policy](https://github.com/openjournals/jose/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSE code of conduct](https://jose.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/JesseBonanno/IndeterminateBeam/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** v2.1.6
- [x] **Authorship:** Has the submitting author (@jessebonanno) made substantial contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation? (and documentation is sufficient?)
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [x] **Installation instructions:** Is there a clearly stated list of dependencies? (Ideally these should be handled with an automated package management solution.)
- [x] **Example usage:** Do the authors include examples of how to use the software?
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSE conflict of interest policy](https://github.com/openjournals/jose/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSE code of conduct](https://jose.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/JesseBonanno/IndeterminateBeam/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** v2.1.6
- [x] **Authorship:** Has the submitting author (@jessebonanno) made substantial contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation? (and documentation is sufficient?)
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [x] **Installation instructions:** Is there a clearly stated list of dependencies? (Ideally these should be handled with an automated package management solution.)
- [x] **Example usage:** Do the authors include examples of how to use the software?
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @amcc1996 , @nicoguaro it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/jose-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/jose-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.26 s (150.7 files/s, 27883.2 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          12            762           1577           2997
HTML                             4              0              0            268
Markdown                         3             84              0            212
Jupyter Notebook                 3              0            533            147
TeX                              2             15              2            147
reStructuredText                 6            131             84            145
DOS Batch                        1              8              1             26
YAML                             1              5             12             16
JSON                             1              0              0             14
make                             1              4              7              9
JavaScript                       1              2              0              6
SVG                              4              0              0              4
-------------------------------------------------------------------------------
SUM:                            39           1011           2216           3991
-------------------------------------------------------------------------------


Statistical information for the repository '1e61a67383564d8c36597956' was
gathered on 2021/04/07.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
JesseBonanno                   125         46263          40871          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
JesseBonanno               5344           11.6          1.6               11.55
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.21105/jose.00065 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.7717/peerj-cs.103 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/jose-papers/jose.00111/jose.00111/10.21105.jose.00111.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/jose-papers/blob/jose.00111/jose.00111/10.21105.jose.00111.pdf) :page_facing_up: :point_left:

**Comment by moorepants** [comment]
@whedon set v2.0.1 as version

**Comment by whedon** [comment]
OK. v2.0.1 is the version.

**Comment by moorepants** [comment]
@JesseBonanno, @amcc1996 and @nicoguaro the review will occur here. Anywhere you see JOSS above, take it as JOSE (as we share the review system with JOSS). But the same guidelines apply. @amcc1996 and @nicoguaro make sure to review the documentation for being a reviewer here: https://openjournals.readthedocs.io/en/jose/. Let me know if you have any questions.

How does three weeks for the first pass for review sound? If not good, please propose a duration that can work for you.

**Comment by nicoguaro** [comment]
> 
> 
> @JesseBonanno, @amcc1996 and @nicoguaro the review will occur here. Anywhere you see JOSS above, take it as JOSE (as we share the review system with JOSS). But the same guidelines apply. @amcc1996 and @nicoguaro make sure to review the documentation for being a reviewer here: https://openjournals.readthedocs.io/en/jose/. Let me know if you have any questions.

I have reviewed in JOSE with you as en editor. Should I re-read the documentation or is it the same?

> How does three weeks for the first pass for review sound? If not good, please propose a duration that can work for you.

It works for me.

**Comment by moorepants** [comment]
> I have reviewed in JOSE with you as en editor. Should I re-read the documentation or is it the same?

Always good to remind yourself. I believe that docs website is new (at least since I last did this).

**Comment by nicoguaro** [comment]
> Always good to remind yourself. I believe that docs website is new (at least since I last did this).

Yup, I will read it :).

**Comment by moorepants** [comment]
Great. Feel free to ask any questions here and we'll set April 28th as a deadline for your first pass on the review. Thank you both for your time and expertise.

**Comment by nicoguaro** [comment]
@JesseBonanno, I followed the installation as recommended in the README file and tried the example in the documentation but did not manage to get the plots. When I ran it in Visual Studio Code I got the following error message

> ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed

**Comment by JesseBonanno** [comment]
> @JesseBonanno, I followed the installation as recommended in the README file and tried the example in the documentation but did not manage to get the plots. When I ran it in Visual Studio Code I got the following error message
> 
> > ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed

@nicoguaro I took a look at the google Golab and Binder online examples and they didn't have this issue. To try resolve the issue I have added in jupyter>=1.0.0 into the requirements.txt which I believe should work, although I am having difficulty recreating this issue at the moment. Please try the installation again and let me know if it works for you.

**Comment by nicoguaro** [comment]
> @nicoguaro I took a look at the google Golab and Binder online examples and they didn't have this issue. To try resolve the issue I have added in jupyter>=1.0.0 into the requirements.txt which I believe should work, although I am having difficulty recreating this issue at the moment. Please try the installation again and let me know if it works for you.

@JesseBonanno, I don't think that work. I am using the suggested method of installation

    pip install indeterminate beams

You would need to create a new version there.

I could try installing from the cloned repository, but I first try to use the instructions verbatim since that is what the users will try.

**Comment by JesseBonanno** [comment]
@nicoguaro oh right, yes i would need to update the version. 

For now could you try pip install jupyter>=1.0.0 alongside the pip install indeterminatebeam so i can verify if this fixes the issue you had since i am having trouble recreating this issue.

If it works i will push that version to 2.0.2 and have whedon update the version.

**Comment by nicoguaro** [comment]
> 
> 
> @nicoguaro oh right, yes i would need to update the version.
> 
> For now could you try pip install jupyter>=1.0.0 alongside the pip install indeterminatebeam so i can verify if this fixes the issue you had since i am having trouble recreating this issue.
> 
> If it works i will push that version to 2.0.2 and have whedon update the version.

Yes, it worked. I ran it from a terminal and it opened two tabs in a browser with the plots.

**Comment by JesseBonanno** [comment]
@whedon set v2.0.2 as version

**Comment by whedon** [comment]
I'm sorry @JesseBonanno, I'm afraid I can't do that. That's something only editors are allowed to do.

**Comment by JesseBonanno** [comment]
I have published v2.0.2 to ammend the vscode jupyter notebook issue as described in the previous few comments.

Can we please change the version to v2.0.2 to capture this?

**Comment by moorepants** [comment]
@whedon set v2.0.2 as version

**Comment by whedon** [comment]
OK. v2.0.2 is the version.

**Comment by nicoguaro** [comment]
@JesseBonanno I tried in Visual Studio code and still have some problems, that are solved after installing nbformat. Your change make the plots works, though. I tried them in plain Python and Jupyter notebook.

Regarding the software, it is giving wrong results. Even the example in the README is wrong since it does not satisfy equilibrium in the vertical direction.

**Comment by JesseBonanno** [comment]
@nicoguaro sorry for not noticing this before the submission, this is obviously a big issue and thank you for noticing it.

I compared the solution to the solution from 1.2.3, and I believe the previous solution was correct.

Previous: https://github.com/JesseBonanno/IndeterminateBeam/blob/v1.2.3/docs/examples/readme_example_internal.png
Current: https://github.com/JesseBonanno/IndeterminateBeam/blob/main/docs/examples/readme_example_internal_HD.png

There seems to be an issue with the interpretation of starting and ending points for the UDL's in the update version 2, which hopefully won't take too long to amend (if the UDL is considered to run the full length then i think the solution would have been correct). I will also try to make my unit tests more robust as this wasn't captured in my unit test.

I will fix this up on the weekend.

**Comment by JesseBonanno** [comment]
> Regarding the software, it is giving wrong results. Even the example in the README is wrong since it does not satisfy equilibrium in the vertical direction.

@nicoguaro I have fixed the issue and updated the package to version v2.0.3.

I have also slightly modified the readme example to use a "UDLV" class type rather than a "DistributedLoadV", since it is more relevant to the example. The original way that the code is written does work though with the new changes, but the DistributedLoad class is for arbritrary loads and runs the example slower than just using UDL class hence the change.

I plan on adding more tests to the unittesting, and plan on updating the package documentation to better reflect changes from moving to version 2 of the software, although i can do this in parallel to the JOSE review.

**Comment by JesseBonanno** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/jose-papers/jose.00111/jose.00111/10.21105.jose.00111.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/jose-papers/blob/jose.00111/jose.00111/10.21105.jose.00111.pdf) :page_facing_up: :point_left:

**Comment by nicoguaro** [comment]
@JesseBonanno, I installed the new version (this time in Linux) and it worked. The example seems to be working as well.

I created and [issue](JesseBonanno/IndeterminateBeam#21) for an example that does not work.

**Comment by JesseBonanno** [comment]
@nicoguaro 

> I installed the new version (this time in Linux) and it worked. The example seems to be working as well.

That's great, thanks.

> I created and [issue](JesseBonanno/IndeterminateBeam#21) for an example that does not work.

I have just fixed and closed the issue. Great spot, thank you for finding it.

**Comment by amcc1996** [comment]
I have just finished my first thorough pass on the repository and paper. With regard to the paper itself, I have no further comments, besides the small remarks and corrections that @JesseBonanno has already applied. Concerning the overall code + repository, I have both some minor and major comments that I would like to discuss.

**Major revision comments** 
1. I cannot run the [README.md example](https://gist.githubusercontent.com/amcc1996/5973b8cb62c8c845d5dce6d006107fa1/raw/841da117a7fefa8b7a806393f10dca11068a1d79/readme_example.py) in my system. I installed the package in a fresh virtual environment, so hopefully, everything is being run with the correct versions. After running the script, the following error surges in sympy:
```
Traceback (most recent call last):
  File "readme_example.py", line 38, in <module>
    beam.analyse()
  File "/home/amcc/repos/tmp/venv/lib/python3.8/site-packages/indeterminatebeam/indeterminatebeam.py", line 682, in analyse
    ans = float(ans)
  File "/home/amcc/repos/tmp/venv/lib/python3.8/site-packages/sympy/core/expr.py", line 358, in __float__
    raise TypeError("can't convert expression to float")
TypeError: can't convert expression to float
```
At the moment, I can't sort out whether the problem is in the code or on my specific setup. This error occurs sporadically in some other examples that I tried, but without any evident pattern. @JesseBonanno, can you please confirm that you can run the example in the link above?

2. I found the physical units adopted for the quantities involved rather odd. For instance, _meters_ are used for lengths, while _millimetre_ are used for the cross-sectional properties, loads are given in _kN_ and the Young modulus in _MPa_. Would it make more sense to use a consistent system of units? For instance, _m-N-Pa_ or _mm-N-MPa_. I must confess that I struggled a bit to compare some analytical solutions to the results of the code, as several corrections to the units are required. If this is current practice on the field, I completely understand the options, but otherwise, it might be a big source of confusion. What do you think @JesseBonanno ?

3. After running some benchmarks against analytical solutions, I have some doubts about the deflection computation. The code is not giving the correct deflection in some problems ([example](https://gist.githubusercontent.com/amcc1996/5973b8cb62c8c845d5dce6d006107fa1/raw/841da117a7fefa8b7a806393f10dca11068a1d79/console.py)). @JesseBonanno  please give a look at this example and the analytical solution for the deflection on the loading point. Maybe I am not seeing it right, but it does some that the computation is incorrect.

4 Another strange behaviour occurs with this [example](https://gist.githubusercontent.com/amcc1996/5973b8cb62c8c845d5dce6d006107fa1/raw/841da117a7fefa8b7a806393f10dca11068a1d79/pinned_beam.py). Apparently, the solver cannot deal with this setup, and gives some infinite reaction forces. Adding to my stupefaction, the online app can solve it seamlessly (even though the deflections seems wrong, see the analytical solution also provided in the script). Please try to run this script @JesseBonanno and assess if my comments are in line with your own observations.


**Minor revision comments** 
1. In the meantime, the version changed from 2.0.2 to 2.0.3. Could you update it, @moorepants ? I don't know if I have the right permissions to do that.
2. The web application that runs the code is truly useful. However, I think there is no mention to it in the README.md page beyond the small badge on the top. Perhaps the repository could profit if there was some short mention in the *Functionality and usage* section. Any thought on this @JesseBonanno ?
3. The way I see it, the CONTRIBUTING.md file already encompasses all the community guidelines with respect to contributing, issue reporting and support. However, maybe it would be helpful to provide a *Support* section directly in the README.md file (or a SUPPORT.md file), where one could find your contact (email) and maybe the references already provided in the documentation. This is a minor suggestion, if you leave it as is, I will be good with it.
4. The code coverage badge does not make justice to the coverage reports with pytest-cov. On my computer, it registered about 75% coverage, instead of the poor 8% that are shown on the homepage!
5. Maybe the repository could also provide the example in the README.md file in a separate file. This would save the readers and users to copy and paste the code by hand to get the first nicely functioning example.
6. The aesthetics of some error messages is somewhat weird. For instance:
![indeterminatebeam-error](https://user-images.githubusercontent.com/43611757/115122181-85a15200-9fae-11eb-9ace-0e5798f419de.png)
This is not a problem at all. Anyway, I felt the need to report it, in the case this was not your original intention, @JesseBonanno .

I apologise for the long comments, which is exactly the opposite that is asked for in the guidelines. If you feel that some of this discussion can be moved to separate issues, please let me know @moorepants, and I will promptly migrate them.

**Comment by JesseBonanno** [comment]
Hi @amcc1996,

First of all thank you for the time you have put into this. Your feedback is constructive and I appreciate it.

Second of all to make sure we are all on the same page, I have addressed issues here and previous issues found by @nicoguaro and have now published a version 2.0.4, which addresses all issues to date. 

If we can move the review to this version (2.0.4) that would be best.

### Major revision comments

1.	I think this might be related to previous issues that I have addressed with @nicoguaro , as I recognise the source of this error message and it is something that he previously picked up on. I don’t recall seeing issues with that particular example and don’t have issues myself trying right now, so maybe you can re-run with v2.0.4 and see if the issue still occurs.
2.	I understand this sentiment, and initially considered having everything in SI units. I have instead tried to keep units aligned with the conventional units specified in my personal experience with engineering in the industry. The objective was both clarity and to avoid typing unnecessary 0’s if possible. I think the current set up is okay, but a third opinion would be good @nicoguaro 
3.	I think the deflection is correct in the program as it currently is. I took a look at your example and I had a different formula for analytical deflection which had the same result as the calculator. 
![image](https://user-images.githubusercontent.com/53243847/115310013-6d992080-a1b0-11eb-8515-f9f7586fa69d.png)

4.	As per issue 1, this is an issue that I believe was addressed [here](https://github.com/JesseBonanno/IndeterminateBeam/issues/22). The reason it works online is the code online is more up to date in line with v2.0.4.

### Minor Comments
1.	Lets change to v2.0.4
2.	Thanks, sounds like a good idea. I have reworded a small section of the README.md (the Statement of Need now Project Purpose) such that it better aligns with the current state of the project. There is a stronger emphasis on the existence of the web-page now.
3.	Makes sense to me. I have added in some references at the bottom of the readme and a small support section as suggested. 
4.	Wow, 75% is a number that sounds more in line with reality than what I am getting. I have been having trouble getting the coverage to work properly and have been trying for longer than I think is worth it anymore. I am just going to remove it now.
5.	The example is currently in a separate file in docs/examples folder. I have linked a file for the ipynb file and the python file in the read me now at the start of the example.
6.	Yep, not my original intention for it to show like that. I was trying to avoid having the code line exceed too many characters. I have slightly modified the code, not sure if there is a better way to do this?

 Old code
![image](https://user-images.githubusercontent.com/53243847/115310146-af29cb80-a1b0-11eb-8e97-8d71bd97b042.png)

New code
![image](https://user-images.githubusercontent.com/53243847/115310188-be107e00-a1b0-11eb-97d3-9d0435136f13.png)

**Comment by nicoguaro** [comment]
I haven't move to the next step in my review, I was first checking the capabilities. Now I think that it is working properly and maybe next weekend will have some comments regarding documentation, testing and code.

Regarding the specific point on units, I have the same comment as @amcc1996 . At first I thought that it might be something related to Hibbeler's book, but there they use different systems of units (usually consistent). I consider that the use of consistent system of units is important. In the case of the software I think that it is even more important because it add [inessential weirdness](http://www.classmatters.org/2006_07/its-not-them.php) to the problem. You are adding a complexity layer that, I think, is not the point being discussed.

**Comment by amcc1996** [comment]
> 1. I think this might be related to previous issues that I have addressed with @nicoguaro , as I recognise the source of this error message and it is something that he previously picked up on. I don’t recall seeing issues with that particular example and don’t have issues myself trying right now, so maybe you can re-run with v2.0.4 and see if the issue still occurs.

Ok that would explain some of my observations. I was running version v2.0.3 on my system the last weekend. I will repeat my tests the next weekend with the new version, thank you for noticing that!

> 3. I think the deflection is correct in the program as it currently is. I took a look at your example and I had a different formula for an analytical deflection which had the same result as the calculator.

Well, it seems that we are using distinct analytical solutions. Perhaps I took the wrong one out of the shelf. I will take another look on my next pass and come back to you right after! Possibly, that is because I was looking at the deflection at the loading point and not the maximum deflection. Anyway, I will test it once again with the new version.


> 5. The example is currently in a separate file in docs/examples folder. I have linked a file for the ipynb file and the python file in the read me now at the start of the example.

My bad, maybe I didn't look hard enough.

> 6. Yep, not my original intention for it to show like that. I was trying to avoid having the code line exceed too many characters. I have slightly modified the code, not sure if there is a better way to do this?

That is great! A much better and ultimately pretty option, in my opinion! I think you can even remove the `+` operator, and Python will handle it seamlessly.

**Comment by moorepants** [comment]
@whedon set v2.0.4 as version

**Comment by whedon** [comment]
OK. v2.0.4 is the version.

**Comment by JesseBonanno** [comment]
Based on the collective feedback I will update the package to use SI units. I am not sure when i will have time to do this but should be done by the weekend and i will make a comment here when it has been changed.

> I think you can even remove the + operator, and Python will handle it seamlessly. 

Thanks for the tip, i will try it out!

**Comment by moorepants** [comment]
> If you feel that some of this discussion can be moved to separate issues, please let me know @moorepants, and I will promptly migrate them.

Discussion is fine here, but please open any issues on the repo that are needed.

**Comment by whedon** [comment]
:wave: @nicoguaro, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @amcc1996 , please update us on how your review is going (this is an automated reminder).

**Comment by amcc1996** [comment]
@JesseBonanno, I repeated by evaluation with the up-to-date version `v2.0.4`. Unfortunately, some of my concerns persist. In the next points, I will summarise the observations resulting from this second pass.

1. I still cannot run the [readme_example.py](https://github.com/JesseBonanno/IndeterminateBeam/blob/main/docs/examples/readme_example.py) in my system. The error occurs in [line 685 of indeterminatebeam.py](https://github.com/JesseBonanno/IndeterminateBeam/blob/82507c1e1495a9f8b0eaadc2d89b63d168c3ea73/indeterminatebeam/indeterminatebeam.py#L685). If I print the value of `var` and `ans` at that point I get `y_0` and `-0.452244897959184*C1 - 0.0194285714285698`, respectively. To my knowledge, it seems that the error is a consequence of the symbolic variable `C1` that is somehow retained up to that point. I can replicate this error on my laptop with Python version `3.8.5`. However, when I run the same example with an older version (3.6, if I recall correctly), there is no error whatsoever. @JesseBonanno what version of Python are you using? Also, do you have any guess why is this behaving this way?

2. It seems that the `UDLV` class is missing from the imports statements of [example_1.py](https://github.com/JesseBonanno/IndeterminateBeam/blob/main/docs/examples/example_1.py). Do you confirm @JesseBonanno ?

3. I can now confirm that the program returns the correct values of deflection in the benchmarks that I ran. The problem was on my side: in the pinned beam example, I was looking at the displacement at the wrong point; in the console beam example there was a sign-related issue in the analytical formula.

In sum, I think nearly all relevant points have been addressed and confirmed. The only exception is the README example which for some unknown reason exhibits unexpected behaviour. In truth, that error is not exclusive to this example, as it has appeared sporadically during my review. Any thoughs on this @JesseBonanno ?

**Comment by nicoguaro** [comment]
> @JesseBonanno, I repeated by evaluation with the up-to-date version `v2.0.4`. Unfortunately, some of my concerns persist. In the next points, I will summarise the observations resulting from this second pass.
> 
>     1. I still cannot run the [readme_example.py](https://github.com/JesseBonanno/IndeterminateBeam/blob/main/docs/examples/readme_example.py) in my system. The error occurs in [line 685 of indeterminatebeam.py](https://github.com/JesseBonanno/IndeterminateBeam/blob/82507c1e1495a9f8b0eaadc2d89b63d168c3ea73/indeterminatebeam/indeterminatebeam.py#L685). If I print the value of `var` and `ans` at that point I get `y_0` and `-0.452244897959184*C1 - 0.0194285714285698`, respectively. To my knowledge, it seems that the error is a consequence of the symbolic variable `C1` that is somehow retained up to that point. I can replicate this error on my laptop with Python version `3.8.5`. However, when I run the same example with an older version (3.6, if I recall correctly), there is no error whatsoever. @JesseBonanno what version of Python are you using? Also, do you have any guess why is this behaving this way?

I am also having problems with that example.  I just installed last version and this is the output I get:

```
Traceback (most recent call last):
  File "/home/nicoguaro/Documents/reviews/readme_example.py", line 16, in <module>
    beam.analyse() 
  File "/home/nicoguaro/Workspace/IndeterminateBeam/indeterminatebeam/indeterminatebeam.py", line 682, in analyse
    ans = float(ans)
  File "/home/nicoguaro/anaconda3/envs/indeterminate_beam/lib/python3.9/site-packages/sympy/core/expr.py", line 358, in __float__
    raise TypeError("can't convert expression to float")
TypeError: can't convert expression to float
```

**Comment by JesseBonanno** [comment]
@amcc1996 @nicoguaro 
Sorry for the slow replies lately, i have been very overwhelmed with work lately and finding it hard to keep up at the moment.

1. In regards to point 1, i have tried running the examples on my computer again and did not observe the issue. I will have to try set up a fresh install to try see what you are seeing. I do have a vague idea of what the error means, it seems like their might be an equation missing that is needed to resolve the problem into a single solution. As to why there might be an equation missing i am not too sure.
2. Fixed, thanks.
3. Okay cheers.

On my end i will work to resolve the following issues:
1. Change units to SI
2. Determine why readme error occurs and fix.

**Comment by moorepants** [comment]
@amcc1996 and @nicoguaro thanks for the careful reviewing up till this point. It seems that @nicoguaro's checklist is mostly complete. @amcc1996 please update yours further if you haven't. I see there are some outstanding issues that @JesseBonanno is working on. @JesseBonanno, once you've made the updates please report back here. I'm going to be away from the internet till Monday, so I'll check on the status then.

**Comment by amcc1996** [comment]
@moorepants for now I have nothing to add besides the weirdly non-functioning example file in README.md. I believe this is the last point that is missing from my checklist. As soon as we discover what is going wrong with this, I will make just a quick check and get back to the discussion.

**Comment by JesseBonanno** [comment]
Hi all,

Finally managed to get a good chunk of free time this sunday afternoon and have managed to adress all our issues to date:

1. Changed all references to refer to base SI units.  Changes occured for several python files, all examples, unit tests, example png snapshots, readme.md and of course paper.md . As there is a lot of references to units throught the code i wouldnt be surprised if i missed any small references to the old units although i have tried my best to cover everything.

2. Readme error. This one took a while to figure out but it seems to be a float precision error with linsolve which only happens with Sympy version 1.8 which was released 22 days ago. I tested on sympy 1.4 which is the minimum version i specified in the requirements and with only that change alone the code worked. I also tested on version 1.7.1 which is the version before 1.8 and the version i personally have installed at the moment and it also worked. I think the easiest solution at the moment would be to have sympy==1.7.1 in the requirements, but i recall reading that forcing a particular version is bad practice since it could make someone down grade in future. Please advise if this solution is adequate.

3. As per an issue created by @nicoguaro i have updated documentation to include all load classes.

**Comment by JesseBonanno** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/jose-papers/jose.00111/jose.00111/10.21105.jose.00111.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/jose-papers/blob/jose.00111/jose.00111/10.21105.jose.00111.pdf) :page_facing_up: :point_left:

**Comment by moorepants** [comment]
> This one took a while to figure out but it seems to be a float precision error with linsolve which only happens with Sympy version 1.8 which was released 22 days ago.

Is this related: https://github.com/sympy/sympy/issues/21162

**Comment by JesseBonanno** [comment]
@moorepants Seems to be the exact issue. If a fix is released for that issue than I believe it would solve the issue observed here.

**Comment by nicoguaro** [comment]
@moorepants, I think that the package is ready from my perspective. 

I have some minor comments that I think that can improve the package but they don't need to be addressed for me to accept:

- I would suggest to create a separate folder for the paper in the repository. The package could grow and the files related to the paper might be a distraction there.
- In the API (docstrings) for vertical and horizontal loads there isn't a note about what direction is positive and negative. I know that this is covered in the Theory section of the documentation but adding it there would cause no harm and might be useful for calls to `help()` or `?` in Jupyter.
- At the beginning of Theory in docs a diagram would make clearer the balance of forces and moments.
- I think that there is an extra `beambending` at the beginning of the Theory section.
- Consider adding more cases to the test suite. Maybe some of the cases raised during this review process.

**Comment by JesseBonanno** [comment]
@nicoguaro I have addressed your latest comments below.
> I would suggest to create a separate folder for the paper in the repository. The package could grow and the files related to the paper might be a distraction there.

I agree and have updated, will need to update repository to https://github.com/JesseBonanno/IndeterminateBeam/paper for whedon to work. If this is an issue can move paper back.

> In the API (docstrings) for vertical and horizontal loads there isn't a note about what direction is positive and negative. I know that this is covered in the Theory section of the documentation but adding it there would cause no harm and might be useful for calls to help() or ? in Jupyter.

Updated.

> At the beginning of Theory in docs a diagram would make clearer the balance of forces and moments.

Noted, will leave as is for the moment.

> I think that there is an extra beambending at the beginning of the Theory section.

Fixed.

> Consider adding more cases to the test suite. Maybe some of the cases raised during this review process.

Agree. I want to add all the examples to the test suite as they cover a broad range of cases that the package covers. I will work on this when I can.

**Comment by moorepants** [comment]
> @moorepants Seems to be the exact issue. If a fix is released for that issue than I believe it would solve the issue observed here.

If you can add a reproducible example to that issue, it could help.

**Comment by moorepants** [comment]
> I think the easiest solution at the moment would be to have sympy==1.7.1 in the requirements, but i recall reading that forcing a particular version is bad practice since it could make someone down grade in future. Please advise if this solution is adequate.

This is an appropriate solution. Or you could do `>=1.4,<1.8` to be more flexible.

**Comment by moorepants** [comment]
@amcc1996 there is one checkbox for the "examples" that you've left unchecked. Is that now resolved?

**Comment by moorepants** [comment]
@JesseBonanno Please make a new software release with the final changes.

**Comment by moorepants** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/jose-papers/jose.00111/jose.00111/10.21105.jose.00111.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/jose-papers/blob/jose.00111/jose.00111/10.21105.jose.00111.pdf) :page_facing_up: :point_left:

**Comment by JesseBonanno** [comment]
> > @moorepants Seems to be the exact issue. If a fix is released for that issue than I believe it would solve the issue observed here.
> 
> If you can add a reproducible example to that issue, it could help.

The readme example works as a reproducible example, is this adequate?

> This is an appropriate solution. Or you could do >=1.4,<1.8 to be more flexible.

Noted, will update.

> @JesseBonanno Please make a new software release with the final changes.

Will do.

**Comment by moorepants** [comment]
> The readme example works as a reproducible example, is this adequate?

It can, but maybe a simple linsolve example with the values that makes it fail is best.

**Comment by JesseBonanno** [comment]
> It can, but maybe a simple linsolve example with the values that makes it fail is best.

Understood, I will get something back to you soon. I will try to find a simple set of equations that raises the error.

**Comment by JesseBonanno** [comment]
@moorepants 

Package has been updated to v2.0.5.

Example of Sympy v1.8 linsolve error occurs is for a 7000 mm simply supported beam with a 1000 N point load at its centre.

equations = [
y_0 + y_7000 + 1000.0, 
7000*y_7000 + 3500000.0,
0.552486187845304*C2,
3867.40331491713*C1 + 0.552486187845304*C2 + 31583793738.4899*y_0 + 3947974217311.23
]
unknowns = [y_0, y_7000, C1, C2]

solution = (1102922357.72339*C1 - 3.37769972052787e+18, -500.0, 1.0*C1, 0)

Interestingly for the same problem (a simply supported beam with a point load at is centre) did not occur for a 2000, 3000, 4000, 5000 , 6000, 8000, 9000, 7, 70, 700 or 70,000 mm long beam which was every other case tested. A summary of the equations and unknowns in the case for the 2000 mm beam is presented below for comparison.

equations = [
y_0 + y_2000 + 1000.0,
2000*y_2000 + 1000000.0,
0.552486187845304*C2,
1104.97237569061*C1 + 0.552486187845304*C2 + 736648250.460405*y_0 + 92081031307.5506
]

unknown = [y_0, y_2000, C1, C2]

solution = (-500.0, -500.0, 250000000.0, 0)

Also worth noting that the example in which we observed this error first the beam length was also 7000 mm. 

Also worth noting that this error does not occur if the Sympy installation is changed to 1.7.1 or 1.4 (other versions have not been tested).

**Comment by amcc1996** [comment]
@moorepants  yes, I left that unchecked because I could not run the example in the documentation. I will look at it today and if goes seamlessly, for me will be ready to go.

**Comment by JesseBonanno** [comment]
Just realised i made a silly mistake when changing the units as I changed to N and mm rather than N and m. I would like to fix this before going forward. May take a few days due to other commitments.

**Comment by moorepants** [comment]
That's no problem. Make our final updates and we'll ensure the reviewers sign off.

**Comment by amcc1996** [comment]
@JesseBonanno the README example now works just fine!
@moorepants , with this I complete my checkbox list. However, I would really appreciate running some examples with the fully corrected version that I believe @JesseBonanno will release soon. Just to make one last check, since some modifications have been introduced relative to my last review.

**Comment by moorepants** [comment]
> However, I would really appreciate running some examples with the fully corrected version that I believe @JesseBonanno will release soon.

Sounds good.

@JesseBonanno, you have all current comments. Please make final adjustments and the reviewers will make a final pass after that.

**Comment by JesseBonanno** [comment]
Hi all,

I have just finished making changes regarding the units. I decided to make the implementation a little bit more flexible since I would prefer to work in alternate units and so the package now uses SI units (N and m) by default but the units used can be changed using the update_units() method. The units available are shown in the units.py file that I have added to the indeterminatebeam folder.

I have updated all examples, example plot png images, paper.md file, readme.md file, testing file, docstrings and read the docs pages to match these changes. Graphs now also explicitly refer to units everywhere, which previously wasn't the case, to remove ambiguity that might be caused by this additional functionality.

The testing file also now includes the readme example and another interpretation of the readme example using different units to provide testing for the update_units() function and method.

I have yet to complete updating the web app to work with the update_units() method but I think it is irrelevant for our purposes, so I can work on it in parallel to the review here.

The update has been released as version v2.1.0.

The changes made in this version are fairly broad and i have made changes to the analysis function so there is certainly potential for error. I have double checked and walked over all the lines in the files myself and everything seems in good order to me. Although this was initially meant to be a quick double check on your behalf I would advise care in review with a moderate review of the examples in the simple_demo.ipynb file and any other examples you like to use. 

Thanks in advance!

**Comment by nicoguaro** [comment]
@JesseBonanno, I think that I will need some time re-review since there is a lot of changes.

Regarding some quick review I did, I have the following comments:

- I think that there a couple of tutorials regarding the update of units and query points are needed in the documentation. For this type of functionality an API entry is not enough, in my opinion.
- The new web app is less intuitive than the one before.
- I found a mistake in the deflection for the query points. I'll open an issue.

**Comment by JesseBonanno** [comment]
@nicoguaro
`I think that I will need some time re-review since there is a lot of changes.`
Understandable, I agree.

`I think that a couple of tutorials regarding the update of units and query points are needed in the documentation. For this type of functionality an API entry is not enough, in my opinion.`
Okay that is a fair consideration, I currently only have one example in the documentation and will add others.

`The new web app is less intuitive than the one before.`
Currently in the process of fixing up the web app, when I am done I hope that it will appear very similiar to before except with the additional settings to play with units. 

`I found a mistake in the deflection for the query points. I'll open an issue.`
This is a rounding issue where the program rounds to 3 dp, which was more so for presentation purposes. I can change the rounding to only occur for plotting purposes.

I also found an issue with Imperial Units not working which I have currently addressed with a few additional lines although I have not uploaded the solution yet.

**Comment by JesseBonanno** [comment]
@nicoguaro I have now updated the web-app to work properly based on the update_unit() method. I have also updated the examples to cover all the examples shown in the jupyter notebook example. 

I am not sure how to approach the rounding issue. I have changed the implementation slightly to make sure no rounding occurs on the backend. Numbers are all displayed using the f string :.3f method for display purposes. To get more precise numbers the user would have to either use the get_deflection() method which would provide no rounding or they would have to change the units to something more sensitive such as mm.

**Comment by amcc1996** [comment]
Hello everybody,

I will only be able to look at this in full detail in the next weekend. As soon as I have some feedback I will come back to you @JesseBonanno .

**Comment by moorepants** [comment]
It seems @JesseBonanno has made changes to address all issues, including making some larger changes in the software as a byproduct of fixing the review comments. We'll need a final pass from both reviewers to ensure the checklists are still valid, as it sounds like some new issues could have been introduced. Is another two weeks a suitable for deadline for the final pass?

**Comment by moorepants** [comment]
@whedon set v2.1.0 as version

**Comment by whedon** [comment]
OK. v2.1.0 is the version.

**Comment by amcc1996** [comment]
@JesseBonanno , I repeated the review process on the updated version (v2.1.0) and did not see a false note anywhere. The possibility to select proper physical units (with the SI system being set as default) is a massive improvement, in my opinion. The documentation was also taken to another level, and the sections pertaining to the treatment of the units is very clear. I ran the unit tests packages with the software and everything was fine. I also looked at all the examples I had created for the first pass and adapted them to the new unit framework and everything looks good!

The only recommendation I have at this moment is to update the paper to refer to the current implementation of the physical units (I believe it still uses mm in the exemplary code snippets in the last version). At first, I was also tempted to advise you to make some comments on the default physical units in the article, as well as the flexibility that the program entails in this regard, but now I am not really sure, as you already refer the reader to the online documentation. Perhaps @nicoguaro could also give some opinion on this.

**Comment by JesseBonanno** [comment]
@amcc1996 Glad to hear you see the value in the unit update, and thanks for the positive feedback.

I updated the paper.md and README.md to have a simple sentence mentioning the change as below:

> The units used throughout the Python package are the base SI units (newtons and metres), but can be updated using the update_units method. Units and load direction conventions are described in the package documentation.

I fixed a minor bug that was found earlier in the review process that I accidentally reintroduced in v2.1.0. 

I also played around with the setup.py file a bit in order to get PyPi to show the readme.md as the description. As a result the version has now jumped ahead to v2.1.4.

**Comment by JesseBonanno** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/jose-papers/jose.00111/jose.00111/10.21105.jose.00111.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/jose-papers/blob/jose.00111/jose.00111/10.21105.jose.00111.pdf) :page_facing_up: :point_left:

**Comment by nicoguaro** [comment]
@moorepants, I think that this is ready to move forward.

**Comment by moorepants** [comment]
@whedon set v2.1.4 as version

**Comment by whedon** [comment]
OK. v2.1.4 is the version.

**Comment by moorepants** [comment]
Ok, great! Thanks for the reminder. I see that all check boxes are done and it seems that all points have been addressed. I need to take a pass through the paper to make sure things are in order for the final publication and then I think we can approve. I'll try to do that tomorrow afternoon.

**Comment by moorepants** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/jose-papers/jose.00111/jose.00111/10.21105.jose.00111.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/jose-papers/blob/jose.00111/jose.00111/10.21105.jose.00111.pdf) :page_facing_up: :point_left:

**Comment by moorepants** [comment]
@JesseBonanno I have checked over everything and it seems your paper can be accepted! Very nice work. You will need to archive your software (e.g. on Zenodo) for the version that will be tied to the paper (currently v2.1.4). Post the archieve DOI here. Once you make the archive I will ask the editor in chief to formally accept and publish the paper.

@nicoguaro and @amcc1996 thank you both very much for the excellent review. This package has many nice improvements from your comments. We greatly appreciate your time and energy.

**Comment by moorepants** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.21105/jose.00065 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.7717/peerj-cs.103 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by moorepants** [comment]
@whedon check repository

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.25 s (228.5 files/s, 115661.1 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          27            962           1994           3832
Jupyter Notebook                 5              0          19079            769
HTML                             5              0              0            335
reStructuredText                 6            289            177            291
Markdown                         4             94              0            245
TeX                              2             15              2            147
DOS Batch                        1              8              1             26
YAML                             1              5              7             19
JSON                             1              0              0             14
make                             1              4              7              9
JavaScript                       1              2              0              6
SVG                              2              0              0              2
-------------------------------------------------------------------------------
SUM:                            56           1379          21267           5695
-------------------------------------------------------------------------------


Statistical information for the repository 'e198a616d73a50f49f1a0b67' was
gathered on 2021/05/28.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
JesseBonanno                   156         49627          42783          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
JesseBonanno               6796           13.7          2.6               12.88
```

**Comment by JesseBonanno** [comment]
> I have checked over everything and it seems your paper can be accepted!

Thats great! Thanks everyone for dedicating their time and knowledge in this review. Everyone made valuable contributions and put in a great deal of effort. I appreciate it very much. @nicoguaro @amcc1996  @moorepants 

> You will need to archive your software (e.g. on Zenodo) for the version that will be tied to the paper (currently v2.1.4).

I had some difficulty with Zenodo generating the upload automatically with new Github releases (v2.1.5, v2.1.6). I have manually uploaded v2.1.6 via Zenodo and generated DOI **10.5281/zenodo.4840592**. There are no real changes between 2.1.4 and 2.1.6 that need to be reviewed, just updates to the version reference.

**Comment by moorepants** [comment]
@whedon set v2.1.6 as version

**Comment by whedon** [comment]
OK. v2.1.6 is the version.

**Comment by moorepants** [comment]
@whedon set 10.5281/zenodo.4840592 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4840592" target="_blank">10.5281/zenodo.4840592</a> is the archive.

**Comment by moorepants** [comment]
@JesseBonanno The DOI for the archive does not seem to resolve: https://doi.org/10.5281/zenodo.4840592

**Comment by JesseBonanno** [comment]
@moorepants yeah there does seem to be something wrong. I have contacted support and hopefully they can help to resolve the issue. The upload page [here](https://zenodo.org/record/4840592/#.YLI-K6gzY2w) seems to show everything in working order and has the DOI, but the DOI does not seem to be activated.

In the mean time i will try an alternative to zenodo.

**Comment by JesseBonanno** [comment]
@moorepants Zenodo is working now with the automatic Github web hook having just been properly processed, the DOI can be changed to 10.5281/zenodo.4851065 for version v2.1.6. The following URL works for me https://doi.org/10.5281/zenodo.4851065

![image](https://user-images.githubusercontent.com/53243847/120097726-831d4500-c175-11eb-957f-3c5adc5adace.png)

**Comment by moorepants** [comment]
@whedon set 10.5281/zenodo.4851065 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4851065" target="_blank">10.5281/zenodo.4851065</a> is the archive.

**Comment by moorepants** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.21105/jose.00065 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.7717/peerj-cs.103 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/jose-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/jose-papers/pull/55

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/jose-papers/pull/55, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by moorepants** [comment]
@labarba This paper is ready!

**Comment by amcc1996** [comment]
Congratulations on your publication @JesseBonanno! Your contribution to open source software and education is very valuable, and I truly believe that both students and engineers can find value in this work.
Thanks for the help and collaboration in the review process @moorepants and @nicoguaro!

**Comment by labarba** [comment]
hi @JesseBonanno 👋  — we do like to request that the metadata of the Zenodo deposit match the paper, in this case the title needs to be manually edited in the metadata. Could you make that tweak and let me know?

**Comment by JesseBonanno** [comment]
Hi @labarba I have just updated the metadata.

I changed the title to be "IndeterminateBeam: A Python package for solving 1D indeterminate beams" and changed the description to be an introductory segment from the paper.

**Comment by labarba** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSE!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/jose-papers/pull/56
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/jose.00111](https://doi.org/10.21105/jose.00111)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by labarba** [comment]
Congratulations, @JesseBonanno, your JOSE paper is published! 🚀 

Huge thanks to our Editor: @moorepants, and the Reviewers: @amcc1996 , @nicoguaro — this adventure is only possible because of your contributions 🙏

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://jose.theoj.org/papers/10.21105/jose.00111/status.svg)](https://doi.org/10.21105/jose.00111)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/jose.00111">
  <img src="https://jose.theoj.org/papers/10.21105/jose.00111/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://jose.theoj.org/papers/10.21105/jose.00111/status.svg
   :target: https://doi.org/10.21105/jose.00111
```
This is how it will look in your documentation:

[![DOI](https://jose.theoj.org/papers/10.21105/jose.00111/status.svg)](https://doi.org/10.21105/jose.00111)

**We need your help!**

Journal of Open Source Education is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future.
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-jose

**Comment by labarba** [comment]
:loudspeaker:  New Reviewer Sign-up Form!
https://forms.gle/Rizd3TcHnQKhrbYY7
